File: 88.merge-sorted-array.java
/*
 * @lc app=leetcode id=88 lang=java
 *
 * [88] Merge Sorted Array
 */

// @lc code=start
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = 0, j = 0;
        int[] tmp = new int[m + n];
        int idx = 0;
        while (i != m && j != n) {
            if (nums1[i] < nums2[j]) {
                tmp[idx++] = nums1[i++];
            } else {
                tmp[idx++] = nums2[j++];
            }
        }
        while (i != m) {
            tmp[idx++] = nums1[i++];
        }
        while (j != n) {
            tmp[idx++] = nums2[j++];
        }
        for (int ix = 0; ix != nums1.length; ++ix) {
            nums1[ix] = tmp[ix];
        }
    }
}
// @lc code=end

File: 88.merge-sorted-array.cpp
/*
 * @lc app=leetcode id=88 lang=cpp
 *
 * [88] Merge Sorted Array
 */

// @lc code=start

#include <vector>

using namespace std;

class Solution {
 public:
  void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
    if (m == 0 && n == 0) return;
    auto iter = nums1.end() - 1;
    int i1 = m - 1, i2 = n - 1;
    while (i1 >= 0 && i2 >= 0) {
      if (nums1[i1] > nums2[i2]) {
        *iter = nums1[i1]; iter -= 1; i1 -= 1;
      } else {
        *iter = nums2[i2]; iter -= 1; i2 -= 1;
      }
    }
    while (i1 >= 0) {
      *iter = nums1[i1]; iter -= 1; i1 -= 1;
    }
    while (i2 >= 0) {
      *iter = nums2[i2]; iter -= 1; i2 -= 1;
    }
  }
};
// @lc code=end
